src/turbine/retransmit_service.zig

```diff
@@ -0,0 +1,430 @@
+
+    var retransmit_sockets: std.ArrayList(UdpSocket) = std.ArrayList(UdpSocket).init(allocator);
+    defer {
+        for (retransmit_sockets.items) |socket| socket.close();
+        retransmit_sockets.deinit();
+    }
+
+    for (0..num_retransmit_sockets) |_| {
+        var socket = try UdpSocket.create(.ipv4, .udp);
+        try socket.bind(try EndPoint.parse("0.0.0.0:0"));
+        try retransmit_sockets.append(try UdpSocket.create(.ipv4, .udp));
+    }
+
+    const receive_thread = try std.Thread.spawn(
+        .{},
+        receiveShreds,
+        .{
+            allocator,
+            my_contact_info,
+            bank_fields,
+            leader_schedule_cache,
+            receiver,
+            &receive_to_retransmit_channel,
+            gossip_table_rw,
+            rand,
+            exit,
+            logger,
+            &stats,
+            overwrite_stake_for_testing,
+        },
+    );
+
+    for (0..num_retransmit_threads) |_| {
+        try retransmit_threads.append(try std.Thread.spawn(
+            .{},
+            retransmitShreds,
+            .{
+                allocator,
+                &receive_to_retransmit_channel,
+                &retransmit_to_socket_channel,
+                &stats,
+                exit,
+            },
+        ));
+    }
+
+    for (retransmit_sockets.items) |socket| {
+        try socket_threads.append(try std.Thread.spawn(
+            .{},
+            socket_utils.sendSocket,
+            .{
+                socket,
+                &retransmit_to_socket_channel,
+                logger,
+                false,
+                exit,
+                {},
+            },
+        ));
+    }
+
+    receive_thread.join();
+    for (retransmit_threads.items) |thread| thread.join();
+    for (socket_threads.items) |thread| thread.join();
+}
+
+/// Receive shreds from the network, deduplicate them, and then package
+/// them into RetransmitShredInfo's to be sent to the retransmit shred threads.
+fn receiveShreds(
+    allocator: std.mem.Allocator,
+    my_contact_info: ThreadSafeContactInfo,
+    bank_fields: *const BankFields,
+    leader_schedule_cache: *LeaderScheduleCache,
+    receiver: *Channel(Packet),
+    sender: *Channel(RetransmitShredInfo),
+    gossip_table_rw: *RwMux(sig.gossip.GossipTable),
+    rand: Random,
+    exit: *AtomicBool,
+    logger: Logger,
+    stats: *Stats,
+    overwrite_stake_for_testing: bool,
+) !void {
+    var turbine_tree_cache = TurbineTreeCache.init(allocator);
+    defer turbine_tree_cache.deinit();
+
+    var deduper = try ShredDeduper(2).init(
+        allocator,
+        rand,
+        DEDUPER_NUM_BITS,
+    );
+    defer deduper.deinit();
+
+    while (!exit.load(.acquire)) {
+        var receive_shreds_timer = try sig.time.Timer.start();
+
+        const receiver_len = receiver.len();
+        if (receiver_len == 0) continue;
+
+        var shreds = try std.ArrayList(Packet).initCapacity(allocator, receiver_len);
```

Comment:
this arraylist can be moved out the loop and re-used (ie, instead of deinit use clearRetainingCapacity)

Reply:
2ee5728a